File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        while left <= right:
            m = left + ((right - left) // 2)

            print("m", m)
            print("left", left)
            print("right", right)
            print("m_val", nums[m])

            if nums[m] == target:
                return m
            if target > nums[m]:
                left = m + 1
            if target < nums[m]:
                right = m - 1
            
        return -1 

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.go
func productExceptSelf(nums []int) []int {
	n := len(nums)
	// prefix slice: value of i is the product of all left elements of i
	prefix := make([]int, n)
	// suffix slice: value of i is the product of all right elements of i
	suffix := make([]int, n)
	// result slice: value of i is product of all except self
	result := make([]int, n)

	prefix[0] = 1
	for i := 1; i < n; i++ {
		prefix[i] = prefix[i-1] * nums[i-1]
	}

	suffix[n-1] = 1
	for i := n - 2; i >= 0; i-- {
		suffix[i] = suffix[i+1] * nums[i+1]
	}

	for i := range result {
		result[i] = prefix[i] * suffix[i]
	}

	return result
}